Mapify/Patches/VisualSwitchPatch.cs

```diff
@@ -0,0 +1,75 @@
+﻿using System.Collections;
+using HarmonyLib;
+using UnityEngine;
+
+namespace Mapify.Patches
+{
+    /// <summary>
+    /// Makes the visual part of the switch work with switches that have more than 2 branches.
+    /// </summary>
+    [HarmonyPatch(typeof(VisualSwitch), nameof(VisualSwitch.PlayAnimation))]
+    public class VisualSwitch_PlayAnimation_Patch
+    {
+        private static readonly int Speed = Animator.StringToHash("speed");
+        private const int ANIMATION_LAYER = 0;
+        private const string STATE_NAME = "junction";
+
+        private static bool Prefix(VisualSwitch __instance)
+        {
+            if (__instance.junction.outBranches.Count <= 2) return true;
+
+            if (!(bool)(Object)__instance.animator) return false;
+
+            __instance.EnableAnimator();
+
+            var previousTime = __instance.animator.GetCurrentAnimatorStateInfo(ANIMATION_LAYER).normalizedTime;
+            var newTime = map(
+                __instance.junction.selectedBranch,
+                0, __instance.junction.outBranches.Count - 1,
+                0, 1
+            );
+
+            var playForward = newTime > previousTime;
+
+            // set animation direction
+            var speed = playForward ? __instance.speedMult : -__instance.speedMult;
+            __instance.animator.SetFloat(Speed, speed);
+
+            // play animation
+            __instance.animator.Play(STATE_NAME, ANIMATION_LAYER, previousTime);
+
+            __instance.StopAllCoroutines();
+            var routine = Pause(newTime, playForward, __instance.animator);
+            __instance.StartCoroutine(routine);
+
+            return false;
+        }
+
+        private static IEnumerator Pause(float pauseTime, bool playForward, Animator animator)
+        {
+            if (playForward)
+            {
+                while (animator.GetCurrentAnimatorStateInfo(ANIMATION_LAYER).normalizedTime < pauseTime)
+                {
+                    yield return null;
+                }
+            }
+            else
+            {
+                while (animator.GetCurrentAnimatorStateInfo(ANIMATION_LAYER).normalizedTime > pauseTime)
+                {
+                    yield return null;
+                }
+            }
+
+            // pause
+            animator.SetFloat(Speed, 0);
+
+            Mapify.Log($"{nameof(Pause)} {animator.GetCurrentAnimatorStateInfo(ANIMATION_LAYER).normalizedTime} PAUSED");
+        }
+
+        private static float map(float value, float in_min, float in_max, float out_min, float out_max) {
+            return (value - in_min) * (out_max - out_min) / (in_max - in_min) + out_min;
+        }
```

Comment:
Do we have this somewhere already? If not, we might want a generic "math utils" type of class so this stuff doesn't get duplicated by accident, should we need it somewhere else in the future.

Reply:
I made a MathUtils class